### PR TITLE
add Dirichlet kernel diric

### PR DESCRIPTION
**Pull request by JeffFessler**
Addresses #400 
It is desirable to have this in a repo because naive implementations based on the Wikipedia definition alone will fail near multiples of 2\pi.

### CODE REVIEW

**Comment by galenlynch** [comment]
Thanks for doing this! I'll review it soon.

**Comment by JeffFessler** [comment]
Hi @galenlynch, just thought I'd ping you about this...

**Comment by rob-luke** [comment]
@JeffFessler could you please add this to the docs.

File: src/diric.jl
Comment (rob-luke) [inline src/diric.jl:2]: These comments are normally not included in the code. Please follow the format from the rest of the repository.

Your contribution will be visible in the git logs. And also in the contributors list generated on the GitHub website.

Comment (JeffFessler) [reply src/diric.jl]: done

**Comment by JeffFessler** [comment]
@rob-luke and @galenlynch, do y'all need some assistance maintaining this package?
I teach DSP and I would really like students to switch from Matlab to Julia, but that really
means it needs to have an actively maintained DSP package.  How can I help?

**Comment by rob-luke** [comment]
> it needs to have an actively maintained DSP package

I think its unfair to accuse this package of being unmaintained. It has had constant attention since 2012, almost 10 years. Please be understanding that the time to review a PR varies based on the complexity of the change and also external factors (I assume most of us are academics and have less time to review during grant season for example, we also have personal lives). The considered approach to merging changes is what has maintained the high quality of this repository.

> How can I help?

You already are by submitting a PR! Other things that can help are reviewing other PRs, we usually try and get two approvals before merging. Similarly, triaging the issues and suggesting to close items that have already been addressed is helpful too. Also as you've already done, a gentle reminder via an @ ping can help to remind people to review can help too 😄

File: src/diric.jl
Comment (rob-luke) [inline src/diric.jl:1]: Just a comment on code organisation for @martinholters to decide on.

Given that this function is documented in `utils` should the code go in `src/utils.jl`? Similarly, for the tests?

**Comment by JeffFessler** [comment]
Sorry, I meant this to be an expression of concern, not an accusation.  I realize that people have a lot going on in their lives especially right now.  So from my perspective when I have a PR that has been there for almost 3 months it leaves the impression that the maintainer's attention is elsewhere (quite understandably) so perhaps it would be a relief to have more hands on deck.  I'll try a simpler @ ping next time.  (I read somewhere else that 2 weeks considered reasonable for sending another ping.)  Thanks for the quick reply.

File: src/diric.jl
Comment (rob-luke) [reply src/diric.jl:1]: Ops I meant @galenlynch sorry

**Comment by rob-luke** [comment]
>  more hands on deck.

Absolutely! Like I said, reviewing PRs and addressing issues is a great help.

File: src/diric.jl
Comment (martinholters) [inline src/diric.jl:28]: This is a rather indirect condition. I'd propose `isapprox(denom, 0; atol=...)` instead, where I'm not sure what to use for the `...` here to be equivalent. (Which is part of the problem I'm having with this as it is now). Actually, there is no value that is exactly equivalent as
```julia
julia> denom = 1.4901161342859267e-8
1.4901161342859267e-8

julia> isapprox(1 + denom, 1)
true

julia> denom = -denom
-1.4901161342859267e-8

julia> isapprox(1 + denom, 1)
false
```
Whereas the proposed alternative would be symmetric around zero (which also seems to make more sense.)

File: src/diric.jl
Comment (martinholters) [inline src/diric.jl:33]: Doesn't `2T(π)` (or `4T(π)`) work here irrespective of `T`?

**Review by martinholters — CHANGES_REQUESTED** [review]
Haven't checked the math, but superficially looks good to me except for a few minor points.

File: src/diric.jl
Comment (JeffFessler) [reply src/diric.jl]: Yes it does; not sure why I though I needed to do that.  Fixed - thanks!

File: src/diric.jl
Comment (JeffFessler) [reply src/diric.jl]: Very good points.  I agree that `atol` is preferable here for symmetry and clarity.
Because `isapprox` has a lot of unnecessary stuff in it here, the updated PR uses an `atol` that is provided as a keyword option with the most sensible default (that stems from `isapprox` code).
It is pretty unlikely that any user will ever use anything other than that default,
but I prefer not to bury constants inside functions...  See what you think about the new version.
Thanks!

File: src/diric.jl
Comment (JeffFessler) [reply src/diric.jl:1]: It's up to you of course, but personally I prefer to keep unrelated methods in different files.

File: src/diric.jl
Comment (martinholters) [inline src/diric.jl:12]:
```suggestion
``1/n \\sum_{k=-(n-1)/2}^{(n-1)/2} e^{i k Ω}``,
```
Alternatively, `e^{i (n-1) Ω/2}/n \\sum_{k=0}^{n-1} e^{i k Ω}` could be used, which would also hold for even `n`.

**Comment by martinholters** [comment]
I've just tried this and it looks like the accuracy is very good around zero, but deteriorates at higher multiples of 2π. I'd therefore propose to unconditionally first do the wrapping of Ω to within [-π, π] for odd `n` and [-2π, 2π] for even `n`.

File: src/diric.jl
Comment (martinholters) [inline src/diric.jl:32]: Do these references actually work in some context? Building the docs locally, I get warnings about them during build and dead links in the output.

File: src/diric.jl
Comment (martinholters) [reply src/diric.jl]: Actually, letting this sink in for a bit, I don't think `atol` should be a parameter. The function should automatically pick a value based on the input type. This is not a parameter one might want to change based on context. Ideally, it should be chosen at the cross-over point where the naive formula becomes less exact than the Taylor approximation.

So I tried to figure out where that is. Around zero, it seems the error of the naive formula should be bounded by `2eps(T)`. (And some numerical experiments seem to verify this.) Given that the correct result is close to one, just the numerical error of evaluating the Taylor approximation might be comparable, so even in a range where the error due to the truncation of the series is small, the Taylor approximation might still not be beneficial. I'm not sure what happens when subnormals come into play, though. But in that case, probably `sin(x)==x`, so the naive formula would give `1.0` which would also probably the closest floating point value. So that indicates that only the case where `Ω/2==0` would require special treatment. A small but non-zero `atol` would feel safer though; `atol=eps(T)` comes to mind. Or we estimate the error of the series truncation by the first omitted term and see where that exceeds `eps(T)`, but that might be overly complicated.

For odd `n`, if we first map Ω to within (-π,π) (as already suggested below), the behavior around zero would all we have to think about. For even `n`, there is also 2π to consider...

**Comment by martinholters** [comment]
Ah, of course, for even `n`, we can also always map to Ω [-π, π]; if that requires shifting by an odd multiple of 2π, we just have to negate the result.

File: src/diric.jl
Comment (martinholters) [inline src/diric.jl:48]: If `ω^2 * (n^2 - 1) / 24 < 0.25*eps(T)` (i.e. `ω < sqrt(6eps(T)/(n^2-1))`), this will evaluate to `one(T)`. So if we fix `atol` at a rather low value as proposed above, it seems we could actually simplify this to just `one(T)` without loosing anything except for really large `n`.

File: src/diric.jl
Comment (JeffFessler) [reply src/diric.jl:32]: Oops, I was following some examples in the base julia docs without fully understanding them and now I realize that probably such refs only work within one package, not between packages.  I will remove them.

File: src/diric.jl
Comment (JeffFessler) [reply src/diric.jl]: Great points!  I agree that any decisions about tolerances should be in the design of the function rather than leaving it to the caller.

File: src/diric.jl
Comment (JeffFessler) [reply src/diric.jl:48]: Very insightful - in my enthusiasm about the Taylor expansion I overlooked that the square will make the 2nd-order term negligible in the precision in hand.  I agree we can simplify to `one`.

File: src/diric.jl
Comment (JeffFessler) [reply src/diric.jl:12]: Doh!  Good catch.

Comment (JeffFessler) [reply src/diric.jl]: BTW I started making other edits before noticing this suggestion.  I did incorporate the suggestion but I'm not enough of a git wizard to know how to "commit" your version.  Sorry about that.

**Comment by JeffFessler** [comment]
@martinholters hopefully I've incorporated all your great suggestions.  I will be traveling the next 2 weeks so I might be slow to reply if you have further refinements.  Thanks!

File: src/diric.jl
Comment (martinholters) [reply src/diric.jl]: No worries.

**Comment by martinholters** [comment]
We're definitely making progress here! I still see problematic numerical errors around 2π for larger even n (e.g. `n=400`). Guess we should either do the "map to [-π, π] and maybe flip the output sign" thing or expand the range where we do the polynomial approximation in that case, where I think I'd prefer the former.

BTW, thanks for your patience here.

File: src/diric.jl
Comment (martinholters) [inline src/diric.jl:44]:
```suggestion
        Ω = rem2pi(Ω, RoundNearest) # [-π,π)
```

File: src/diric.jl
Comment (martinholters) [inline src/diric.jl:46]:
```suggestion
        Ω = 2 * rem2pi(Ω/2, RoundNearest) # [-2π,2π)
```

**Comment by JeffFessler** [comment]
@martinholters, I still had trouble getting good precision near 2\pi for even `n` even after modifying the wrapping in the way I think you were suggesting.  I ended up just doubling the precision for intermediate calculations for the case of even `n` and that seemed to solve it for the tests I was using.  Odd `n` is the most useful case, so the extra "work" of doubling the precision for the even `n` case seems reasonable to me in the interest of providing an accurate result.  Do you have any ideas for a simpler way?

File: src/diric.jl
Comment (martinholters) [inline src/diric.jl:48]:
```suggestion
        Ω = 2 * rem2pi(abs(Ω)/2, RoundNearest) # [-2π,2π)
```

File: src/diric.jl
Comment (martinholters) [inline src/diric.jl:51]:
```suggestion
            Ω -= 2π # (-π,0]
```

File: src/diric.jl
Comment (martinholters) [inline src/diric.jl:52]:
```suggestion
        else if Ω < π
            sign = -1
            Ω += 2π # (0,π]
        end
```

**Review by martinholters — COMMENTED** [review]
This was almost what I had in mind, but missed half of it :smile:

File: src/diric.jl
Comment (JeffFessler) [reply src/diric.jl:48]: Note the `abs` above.  I did this so that I would not need the `else if` suggested below.
And I am pretty sure I tested accuracy near -2\pi to verify.

**Comment by JeffFessler** [comment]
So the Julia 1.0 tests are failing because `BigFloat` in days of old does not support the `precision` kw arg.  Ugh.
That was added in 1.1
https://docs.julialang.org/en/v1/base/numbers/#BigFloats-and-BigInts
The latest commit intends to address it - I didn't not test it locally so someone please run CI...

File: src/diric.jl
Comment (martinholters) [reply src/diric.jl:48]: But e.g.
```julia
julia> 2*rem2pi(4.0, RoundNearest)
-4.566370614359173
```
even though `4.0 > 0`.

**Comment by martinholters** [comment]
> So the Julia 1.0 tests are failing because BigFloat in days of old does not support the precision kw arg. Ugh.

I'd still hope we can get away without this conversion once the "avoid 2π" thing is done correctly.

**Comment by JeffFessler** [comment]
Yep, after fixing up 2\pi cases properly we no longer need BigFloat.
Hopefully it is ready now.  Thanks for your patience!

File: test/diric.jl
Comment (martinholters) [inline test/diric.jl:11]:
```suggestion
    @test @inferred(diric(0, 4)) ≈ 1
    @test @inferred(diric(0 // 1, 5)) == 1
    @test @inferred(diric(4π, 4)) ≈ 1
    @test @inferred(diric(2π, 4)) ≈ -1
    @test @inferred(diric(0, 5)) ≈ 1
    @test @inferred(diric(2π, 5)) ≈ 1
    @test @inferred(diric(π, 5)) == 1 // 5
```
Otherwise, this verifies that the comparison can be inferred.

File: src/diric.jl
Comment (martinholters) [inline src/diric.jl:79]: This is not generally correct? For even `n`, the result should be zero, for `mod(n,4)==3`, it should be `-1 // n`.

File: test/diric.jl
Comment (martinholters) [reply test/diric.jl:11]: Also, can we test a few arbitrary values which are not multiples of 2π?

**Review by martinholters — COMMENTED** [review]
Alright, I think we're getting somewhere.

File: src/diric.jl
Comment (martinholters) [inline src/diric.jl:74]:
```suggestion
diric(Ω::Real, n::Int) = diric(float(Ω), n::Int)
```
Seems to be more typical, though I don't know what this entails exactly.

Comment (JeffFessler) [reply src/diric.jl:68]: Makes sense - probably gives Float32 on a 32-bit cpu...

File: src/diric.jl
Comment (JeffFessler) [reply src/diric.jl:73]: Doh!  I am just going to change it to let `float` take care of it.

File: src/diric.jl
Comment (JeffFessler) [reply src/diric.jl]: Oops, agreed.

File: src/diric.jl
Comment (JeffFessler) [reply src/diric.jl]: Agreed.

**Comment by JeffFessler** [comment]
Hopefully I covered all your suggestions.  In addition, I added a test for accuracy near 2\pi both to verify some more non multiple of pi values and so that if someone comes around later and tries to "simplify" the code to make it look like the matlab or scipy version they will see the accuracy issue.

**Review by martinholters — APPROVED** [review]
Generally looks good to me now. Some minor nits (I feel like I discover something new every time I look), but those are not mandatory in my eyes. We could also merge as is and only iterate more on this if the need actually arises.

File: src/diric.jl
Comment (JeffFessler) [reply src/diric.jl:57]: Yeah, this function is unlikely to be used in any tight inner loop so a bit of corner case type instability seems fine.

**Comment by JeffFessler** [comment]
I went ahead and used `Integer`.  It's fine with me to fully refine it now while it's fresh on our minds...

File: src/diric.jl
Comment (JeffFessler) [reply src/diric.jl]: Actually I don't think a specialized method is worth it now.  pi is special, but 2pi becomes a Float64 and I don't see any reason to give one integer multiple of pi special treatment.  `sin(pi)` does not return 0 in base...

File: src/diric.jl
Comment (martinholters) [reply src/diric.jl:1]: As I've commented a lot on this PR but not on this particular question it might be obvious, but just so it be said: I don't care.

**Review by martinholters — APPROVED** [review]
Good to merge from my side, assuming CI comes out green. I'd like to give it another 24 hours or so in case I spot something else with fresh eyes or someone else wants to chime in, but please bump if I forget to merge then.

File: src/diric.jl
Comment (martinholters) [reply src/diric.jl]: Yeah, there would be `sinpi` for that. Speaking of which... nah, let's save `diricpi` for another day :smile:

**Pull request merged**
